### PR TITLE
Introduce verification checks on subtree_size

### PR DESCRIPTION
**Pull request by jonmeow**
This switches the Verify method to walk postorder so that we can see how much subtree_size is really used, and shift towards removing it. It also starts calling Verify.

Also, I think I'd lost the reserve/size check during Parser refactoring, so I'm putting that back in as part of Verify.

### CODE REVIEW

File: toolchain/parser/parse_node_kind.def [hunk 1]
```diff
@@ -9,78 +9,99 @@
 // desired output. The x-macro for this header is `CARBON_PARSE_NODE_KIND`. The
 // definition provided will be removed at the end of this file to clean up.
 
-#ifndef CARBON_PARSE_NODE_KIND
-#error "Must define the x-macro to use this file."
+#if !(defined(CARBON_PARSE_NODE_KIND) ||          \
+      (defined(CARBON_PARSE_NODE_KIND_BRACKET) && \
+       defined(CARBON_PARSE_NODE_KIND_CHILD_COUNT)))
+#error "Must define the CARBON_PARSE_NODE_KIND x-macro to use this file."
```
Comment (chandlerc) [inline toolchain/parser/parse_node_kind.def]: This error doesn't seem accurate any more?

File: toolchain/parser/parse_node_kind.def [hunk 2]
```diff
@@ -9,78 +9,99 @@
...
 
+// The BRACKET and CHILD_COUNT macros will use CARBON_PARSE_NODE_KIND by default
+// when undefined.
+#ifndef CARBON_PARSE_NODE_KIND_BRACKET
+#define CARBON_PARSE_NODE_KIND_BRACKET(Name, ...) CARBON_PARSE_NODE_KIND(Name)
+#endif
+#ifndef CARBON_PARSE_NODE_KIND_CHILD_COUNT
+#define CARBON_PARSE_NODE_KIND_CHILD_COUNT(Name, ...) \
+  CARBON_PARSE_NODE_KIND(Name)
+#endif
```
Comment (chandlerc) [inline toolchain/parser/parse_node_kind.def]: Would be good to document how these are expected to be used below, what the semantics of the parameters are, etc. I think I can guess from reading, but seems good to be explicit in a comment.

File: toolchain/parser/parse_tree.cpp [hunk 3]
```diff
@@ -188,68 +188,84 @@ auto ParseTree::Print(llvm::raw_ostream& output, bool preorder) const -> void {
...
-    // We're going to pop off some levels of the tree. Check each ancestor to
-    // make sure the offsets are correct.
-    int next_index = n.index() - 1;
-    while (!ancestors.empty()) {
-      ParseTree::Node parent_n = ancestors.back();
-      if ((parent_n.index() - node_impls_[parent_n.index()].subtree_size) !=
-          next_index) {
-        break;
+    } else if (n_impl.kind.child_count() == -2) {
+      // TODO: Allows -2 to mean "unsupported"; see parse_node_kind.def.
```
Comment (chandlerc) [inline toolchain/parser/parse_tree.cpp]: Maybe just have a quick named constant in the kind type to make the meaning of -2 more obvious?

**Review by chandlerc — CHANGES_REQUESTED** [review]
Verify code makes lots of sense, but some adjustments that would help me on the node kind side. I like the direction though. =]

File: toolchain/parser/parse_node_kind.def (hunk 1 above)
Comment (jonmeow) [reply toolchain/parser/parse_node_kind.def]: Adjusted message, does the rephrasing work for you? I'm trying to keep it short to fit on one line for #error

File: toolchain/parser/parse_node_kind.def (hunk 2 above)
Comment (jonmeow) [reply toolchain/parser/parse_node_kind.def]: Added comments to file

File: toolchain/parser/parse_tree.cpp (hunk 3 above)
Comment (jonmeow) [reply toolchain/parser/parse_tree.cpp]: Added TodoFixParseNode

File: toolchain/parser/parse_node_kind.def [hunk 4]
```diff
@@ -1,86 +1,118 @@
 // Part of the Carbon Language project, under the Apache License v2.0 with LLVM
 // Exceptions. See /LICENSE for license information.
-// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-License-Identifier: ApacheTodoFixParseNode.0 WITH LLVM-exception
```
Comment (chandlerc) [inline toolchain/parser/parse_node_kind.def]: Woops... =D
```suggestion
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
```

File: toolchain/parser/parse_node_kind.h [hunk 5]
```diff
@@ -64,6 +67,16 @@ class ParseNodeKind {
...
   void Print(llvm::raw_ostream& out) const { out << name(); }
 
+  // Returns true if the node is bracketed; otherwise, child_count is used.
+  auto has_bracket() const -> bool;
+
+  // Returns the bracketing character for the node. Requires that has_bracket is
+  // true.
+  auto bracket() const -> ParseNodeKind;
+
+  // Returns the child_count for the node. Requires that has_bracket is false.
```
Comment (chandlerc) [inline toolchain/parser/parse_node_kind.h]: Might be worth some description of what this means? Maybe can even refer to it from the x-macro file if you want.

**Review by chandlerc — APPROVED** [review]
Awesome, LG! Typo issue and an optional comment below.

File: toolchain/parser/parse_node_kind.h (hunk 5 above)
Comment (jonmeow) [reply toolchain/parser/parse_node_kind.h]: Iterated on comments.

**Pull request merged**